### ISSUE TITLE
[Bug]: Budget Limiter Memory Cache Sync Can Read Stale Redis Data

### ISSUE THREAD

**Issue by emerzon** [BerriAI/litellm#20886]
### Check for existing issues

- [x] I have searched the existing issues and checked that my issue is not a duplicate.

### What happened?

The `RouterBudgetLimiting._push_in_memory_increments_to_redis()` method uses `asyncio.create_task()` to schedule Redis increment pipelines without awaiting their completion. The `_sync_in_memory_spend_with_redis()` method then immediately fetches from Redis after awaiting `_push_in_memory_increments_to_redis()`, but the pipeline may not have executed yet, causing memory cache synchronization to use stale data. This leads to budget tracking inconsistencies and potential budget overruns in production proxy deployments with multiple concurrent requests.

**Severity:** 9/10 - Regularly visible malfunction in multi-instance or high-concurrency environments. Budget limits can be exceeded silently due to temporary memory cache staleness between periodic sync operations.

## Root Cause & Impact Flow

- **File(s):**
  - `litellm/router_strategy/budget_limiter.py` (lines 498-502, 530)
  
- **Function/Method:**
  - `_push_in_memory_increments_to_redis()` at lines 481-509
  - Line 498-502 creates unawaited task
  
- **Trigger path:**
  ```
  Multiple concurrent API requests increment budget counters
    → RouterBudgetLimiting.async_log_success_event()
      → _increment_spend_in_current_window() 
        → dual_cache.in_memory_cache.async_increment() [adds to memory]
        → redis_increment_operation_queue.append() [queues for Redis]
          
  Periodic sync (every 1 second):
    → periodic_sync_in_memory_spend_with_redis()
      → await _sync_in_memory_spend_with_redis()
        → await _push_in_memory_increments_to_redis()
          → asyncio.create_task(async_increment_pipeline(...)) [NOT AWAITED!]
            Task scheduled but returns immediately
          → Queue cleared
        → Fetch from Redis: async_batch_get_cache() 
          **BUG:** Pipeline hasn't executed yet, fetching stale data!
        → Update memory with stale value
        
  Subsequent budget check:
    → dual_cache.async_get_cache() reads MEMORY first
      → Gets stale value (lost increments!)
      → Budget check passes incorrectly
  ```

- **What goes wrong:** 
  - `asyncio.create_task()` at line 498 schedules Redis pipeline but returns immediately
  - Method completes, but pipeline still pending in event loop
  - `_sync_in_memory_spend_with_redis()` fetches from Redis expecting recent data
  - Gets stale value because pipeline hasn't executed yet
  - Memory cache is overwritten with this stale value

- **Why it's severe in practice:**
  
  1. **Multi-instance deployments:** Proxy runs with multiple workers, each maintaining separate memory caches synced via Redis
  
  2. **Budget enforcement is critical safety mechanism**:
     - `model_max_budget_limiter.py` tracks spend via `_increment_spend_for_key()` 
     - Budget limits prevent revenue loss from exceeded API quotas
     - If budget tracking loses data, users can exceed allocated budgets
   
  3. **Silent budget overrun:** Memory staleness allows requests based on outdated values
   
  4. **Worst-case scenario (demonstrated below):**
     - Memory shows $160 actual spend
     - Redis pipeline hasn't completed yet
     - Memory synced to stale Redis value ($100)
     - Lost $60 (37.5% data loss!)
     - User can overspend by $60 before next sync

### Steps to Reproduce

Test Script: Budget Sync Staleness Demonstration

```python
import asyncio
import sys

from litellm.router_strategy.budget_limiter import RouterBudgetLimiting
from litellm.caching.caching import DualCache


async def test():
    """Demonstrate budget limiter memory sync bug"""

    # Track values separately for Redis and Memory to show the bug
    redis_values = {"provider_spend:openai:1d": 100.0}

    class MockRedis:
        async def async_get_cache(self, key, **kwargs):
            return redis_values.get(key)

        async def async_batch_get_cache(self, key_list, **kwargs):
            return {k: redis_values.get(k) for k in key_list}

        async def async_set_cache(self, key, value, **kwargs):
            redis_values[key] = float(value)

        async def async_increment_pipeline(self, increment_list, **kwargs):
            # Simulate 300ms network delay
            await asyncio.sleep(0.3)
            for op in increment_list:
                current = redis_values.get(op["key"], 0)
                new_value = current + op["increment_value"]
                redis_values[op["key"]] = new_value

    class MockMemory:
        def __init__(self):
            self.memory_values = {"provider_spend:openai:1d": 100.0}

        async def async_get_cache(self, key, **kwargs):
            return self.memory_values.get(key)

        async def async_set_cache(self, key, value, **kwargs):
            print(f"    → Memory cache updated ({key}) to {value}")
            self.memory_values[key] = float(value)

        async def async_increment(self, key, value, **kwargs):
            current = self.memory_values.get(key, 0.0) or 0.0
            new_value = float(current) + float(value)
            await self.async_set_cache(key=key, value=new_value, **kwargs)
            return new_value

    mock_redis = MockRedis()
    mock_memory = MockMemory()
    dual_cache = DualCache(in_memory_cache=mock_memory, redis_cache=mock_redis)

    router_budget = RouterBudgetLimiting(
        dual_cache=dual_cache,
        provider_budget_config={"openai": {"budget_limit": 500.0, "time_period": "1d"}},
    )

    # Simulate two rapid requests on the real RouterBudgetLimiting path:
    # this updates memory AND appends Redis increment ops to the queue.
    await router_budget._increment_spend_in_current_window(
        spend_key="provider_spend:openai:1d",
        response_cost=25.0,
        ttl=86400,
    )
    await router_budget._increment_spend_in_current_window(
        spend_key="provider_spend:openai:1d",
        response_cost=35.0,
        ttl=86400,
    )

    print(
        f"Memory after increments: ${mock_memory.memory_values['provider_spend:openai:1d']}"
    )  # $160
    print(
        f"Redis still shows: ${redis_values.get('provider_spend:openai:1d')}"
    )  # $100 (pending sync)

    # Trigger periodic sync - THIS IS WHERE BUG MANIFESTS!
    await router_budget._sync_in_memory_spend_with_redis()

    # Give scheduled Redis pipeline task time to finish, so we can compare eventual Redis value.
    await asyncio.sleep(0.35)

    final_redis = redis_values.get("provider_spend:openai:1d")
    final_memory = mock_memory.memory_values.get("provider_spend:openai:1d")

    print(f"\nAfter sync:")
    print(f"  Redis (eventual): ${final_redis}")
    print(f"  Memory (stale): ${final_memory}")

    if final_memory < final_redis:
        loss = final_redis - final_memory
        loss_pct = (loss / final_redis) * 100
        print(f"\n❌ BUG CONFIRMED!")
        print(f"   Lost ${loss} ({loss_pct:.1f}% data loss)")
        return False
    else:
        print("\n✓ No bug")
        return True


result = asyncio.run(test())
sys.exit(0 if result else 1)

```

### Observed Result (Before Fix)

```bash
    → Memory cache updated (provider_spend:openai:1d) to 125.0
    → Memory cache updated (provider_spend:openai:1d) to 160.0
Memory after increments: $160.0
Redis still shows: $100.0
    → Memory cache updated (provider_spend:openai:1d) to 100.0
    → Memory cache updated (provider_spend:openai:1d) to 100.0
    → Memory cache updated (provider_budget_start_time:openai) to 1770750694.843928

After sync:
  Redis (eventual): $160.0
  Memory (stale): $100.0

❌ BUG CONFIRMED!
   Lost $60.0 (37.5% data loss)

```

### Expected Result

- After periodic sync, memory cache should match Redis exactly ($160)
- No increments should be lost during the sync process
- Budget tracking should be accurate at all times

### Relevant log output

```shell

```

### What part of LiteLLM is this about?

Proxy

### What LiteLLM version are you on ?

v1.81.8
